webapp/packages/plugin-data-spreadsheet-new/src/DataGrid/DataGridSelection/useGridSelectionContext.tsx

```diff
@@ -35,72 +35,114 @@ export function useGridSelectionContext(modelResultData: IDatabaseDataResult | n
+    const left = Math.min(idx, lastSelectedCell.idx);
+    const right = Math.max(idx, lastSelectedCell.idx);
+    const top = Math.min(rowIdx, lastSelectedCell.rowIdx);
+    const bottom = Math.max(rowIdx, lastSelectedCell.rowIdx);
 
-    selectedCells.clear();
+    const rowSelection = [];
+    for (let colIdx = left; colIdx <= right; colIdx++) {
+      rowSelection.push(colIdx);
+    }
+    for (let rowIdx = top; rowIdx <= bottom; rowIdx++) {
+      selectedCells.set(rowIdx, rowSelection);
+    }
+  }), [selectedCells, lastSelectedCell]);
 
-    for (let column = 0; column <= Math.abs(columnRange); column++) {
-      const idx = isColumnToBottom ? column + currentSelectedCell.idx : currentSelectedCell.idx - column;
-      if (idx === 0) {
-        continue;
-      }
-      for (let row = 0; row <= Math.abs(rowRange); row++) {
-        const rowIdx = isRowToBottom ? row + currentSelectedCell.rowIdx : currentSelectedCell.rowIdx - row;
+  const isSelected = useCallback((idx: number, rowIdx: number) => {
+    if (!selectedCells.has(rowIdx)) {
+      return false;
+    }
 
-        const hash = getHash(idx, rowIdx);
-        selectedCells.set(hash, { idx, rowIdx });
-      }
+    const rowSelection = selectedCells.get(rowIdx)!;
+    return rowSelection.indexOf(idx) !== -1;
+  },
+  [selectedCells]);
+
+  const unSelect = useCallback((idx: number, rowIdx: number) => {
+    if (!selectedCells.has(rowIdx)) {
+      return false;
+    }
+
+    const rowSelection = selectedCells.get(rowIdx)!;
+    const targetIndex = rowSelection.indexOf(idx);
+
+    if (targetIndex !== -1) {
+      rowSelection.splice(targetIndex, 1);
+      selectedCells.set(rowIdx, rowSelection);
+      return true;
     }
-  }), [selectedCells, currentSelectedCell]);
+    return false;
+  }, [selectedCells]);
+
+  const selectCell = useCallback((idx: number, rowIdx: number) => {
+    const rowSelection = selectedCells.get(rowIdx);
+    if (rowSelection === undefined) {
+      return;
+    }
+
+    rowSelection.push(idx);
+    selectedCells.set(rowIdx, rowSelection);
+  }, [selectedCells]);
 
   const select = useCallback((idx: number, rowIdx: number, multiple: boolean, range: boolean) => {
-    setCurrentSelectedCell({ idx: idx === 0 ? idx + 1 : idx, rowIdx });
+    setLastSelectedCell({ idx: idx === 0 ? idx + 1 : idx, rowIdx });
     if (selectedCells.size > 0 && range) {
-      selectRange(idx, rowIdx);
+      selectRange(idx === 0 ? idx + 1 : idx, rowIdx);
       return;
     }
 
     // it's index column
     if (idx === 0) {
-      selectRow(rowIdx, multiple);
+      multiple && isRowSelected(rowIdx) ? unSelectRow(rowIdx) : selectRow(rowIdx, multiple);
       return;
     }
 
     if (!multiple) {
       selectedCells.clear();
     }
 
-    const hash = getHash(idx, rowIdx);
-
-    if (selectedCells.has(hash)) {
-      selectedCells.delete(hash);
+    if (!selectedCells.has(rowIdx)) {
+      selectedCells.set(rowIdx, [idx]);
       return;
     }
```

Comment:
should be replaced with `selectCell` call

Reply:
so seems you should just remove this lines and create new selected row in `selectCell` like I described above